core/src/main/java/com/linecorp/armeria/common/stream/TimeoutSubscriber.java

```diff
@@ -29,70 +29,88 @@
 import io.netty.util.concurrent.ScheduledFuture;
 
 public class TimeoutSubscriber<T> implements Subscriber<T> {
-
+    private static final String TIMEOUT_MESSAGE = "Stream timed out after %d ms (timeout mode: %s)";
     private final Subscriber<? super T> delegate;
     private final EventExecutor executor;
-
     private final StreamTimeoutMode timeoutMode;
+    private final Duration timeoutDuration;
     private ScheduledFuture<?> timeoutFuture;
-
+    private Runnable timeoutTask;
     private Subscription subscription;
-    private final Duration timeoutDuration;
+    private long timeoutNanos;
+    private long lastOnNextTimeNanos;
 
     public TimeoutSubscriber(Subscriber<? super T> delegate, EventExecutor executor, Duration timeoutDuration, StreamTimeoutMode timeoutMode) {
         this.delegate = requireNonNull(delegate, "delegate");
         this.executor = requireNonNull(executor, "executor");
         this.timeoutDuration = requireNonNull(timeoutDuration, "timeoutDuration");
+        timeoutNanos = timeoutDuration.toNanos();
         this.timeoutMode = requireNonNull(timeoutMode, "timeoutMode");
+        timeoutTask = createTimeoutTask();
     }
 
-    private ScheduledFuture<?> scheduleTimeout() {
-        return executor.schedule(() -> {
+    private Runnable createTimeoutTask() {
+        return () -> {
+            if(timeoutMode == StreamTimeoutMode.UNTIL_NEXT) {
+                long currentTimeNanos = System.nanoTime();
+                long elapsedNanos = currentTimeNanos - lastOnNextTimeNanos;
+
+                if(elapsedNanos <= timeoutNanos) {
```

Comment:
Perhaps we should use `<` instead of `<=`? If `elapsedNanos == timeoutNanos`, it means it's been timed out.